ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineForkchoiceUpdated.java

```diff
@@ -68,93 +71,138 @@ public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext)
+      return new JsonRpcErrorResponse(requestId, JsonRpcError.INVALID_FORKCHOICE_STATE);
+    }
 
-    Optional<BlockHeader> finalizedHead =
-        maybeFinalizedHash.flatMap(protocolContext.getBlockchain()::getBlockHeader);
+    // TODO: post-merge cleanup, this should be unnecessary after merge
+    if (!mergeCoordinator.latestValidAncestorDescendsFromTerminal(newHead.get())) {
+      return new JsonRpcSuccessResponse(
+          requestId,
+          new EngineUpdateForkchoiceResult(
+              INVALID_TERMINAL_BLOCK,
+              null,
+              null,
+              Optional.of(newHead.get() + " did not descend from terminal block")));
+    }
 
-    if (newHead.isPresent() && (maybeFinalizedHash.isEmpty() || finalizedHead.isPresent())) {
+    ForkchoiceResult result =
+        mergeCoordinator.updateForkChoice(
+            newHead.get(), forkChoice.getFinalizedBlockHash(), forkChoice.getSafeBlockHash());
+
+    if (result.isFailed()) {
+      final Optional<Hash> latestValid = result.getLatestValid();
+      return new JsonRpcSuccessResponse(
+          requestId,
+          new EngineUpdateForkchoiceResult(
+              INVALID,
+              latestValid.isPresent() ? latestValid.get() : null,
+              null,
+              result.getErrorMessage()));
+    }
 
-      // TODO: post-merge cleanup, this should be unnecessary after merge
-      if (!mergeCoordinator.latestValidAncestorDescendsFromTerminal(newHead.get())) {
-        return new JsonRpcSuccessResponse(
-            requestContext.getRequest().getId(),
-            new EngineUpdateForkchoiceResult(
-                INVALID_TERMINAL_BLOCK,
-                null,
-                null,
-                Optional.of(newHead.get() + " did not descend from terminal block")));
+    // begin preparing a block if we have a non-empty payload attributes param
+    Optional<PayloadIdentifier> payloadId =
+        optionalPayloadAttributes.map(
+            payloadAttributes ->
+                mergeCoordinator.preparePayload(
+                    newHead.get(),
+                    payloadAttributes.getTimestamp(),
+                    payloadAttributes.getPrevRandao(),
+                    payloadAttributes.getSuggestedFeeRecipient()));
+
+    payloadId.ifPresent(
+        pid ->
+            debugLambda(
+                LOG,
+                "returning identifier {} for requested payload {}",
+                pid::toHexString,
+                () -> optionalPayloadAttributes.map(EnginePayloadAttributesParameter::serialize)));
+
+    return new JsonRpcSuccessResponse(
+        requestId,
+        new EngineUpdateForkchoiceResult(
+            VALID,
+            result.getNewHead().map(BlockHeader::getHash).orElse(null),
+            payloadId.orElse(null),
+            Optional.empty()));
+  }
+
+  private boolean isValidForkchoiceState(
+      final Hash safeBlockHash, final Hash finalizedBlockHash, final BlockHeader newBlock) {
```

Comment:
I'd rather see this logic should be in MergeCoordinator.updateForkChoice, where a failure returns a specific ForkchoiceResult that can indicate INVALID_FORK_CHOICE_STATE.  Otherwise the only place we are doing these checks in in a json rpc method, and the public updateForkChoice method is unsafe.

Reply:
As discussed, we can address this in a subsequent PR to keep the size/scope of this PR more reviewable